shell/platform/darwin/ios/framework/Source/FlutterMetalLayer.mm

```diff
@@ -0,0 +1,271 @@
+    (id)kIOSurfaceHeight : @(_drawableSize.height),
+    (id)kIOSurfacePixelFormat : @(pixelFormat),
+    (id)kIOSurfaceBytesPerElement : @(bytesPerElement),
+    (id)kIOSurfaceBytesPerRow : @(bytesPerRow),
+    (id)kIOSurfaceAllocSize : @(totalBytes),
+  };
+
+  IOSurfaceRef res = IOSurfaceCreate((CFDictionaryRef)options);
+  if (self.colorspace != nil) {
+    CFStringRef name = CGColorSpaceGetName(self.colorspace);
+    IOSurfaceSetValue(res, CFSTR("IOSurfaceColorSpace"), name);
+  } else {
+    IOSurfaceSetValue(res, CFSTR("IOSurfaceColorSpace"), kCGColorSpaceSRGB);
+  }
+  return (__bridge_transfer IOSurface*)res;
+}
+
+- (id<CAMetalDrawable>)nextDrawable {
+  @synchronized(self) {
+    // IOSurface.isInUse to determine when compositor is done with the surface.
+    // That a rather blunt instrument and we are probably waiting longer than
+    // we really need to. With triple buffering at 120Hz that results in about
+    // 2-3 milliseconds wait time at beginning of display link callback.
+    // With four buffers this number gets close to zero.
+    if (_totalDrawables < 4) {
+      ++_totalDrawables;
+      IOSurface* surface = [self createIOSurface];
+      MTLTextureDescriptor* textureDescriptor =
+          [MTLTextureDescriptor texture2DDescriptorWithPixelFormat:_pixelFormat
+                                                             width:_drawableSize.width
+                                                            height:_drawableSize.height
+                                                         mipmapped:NO];
+      textureDescriptor.usage =
+          MTLTextureUsageShaderRead | MTLTextureUsageRenderTarget | MTLTextureUsageShaderWrite;
+      id<MTLTexture> texture = [self.device newTextureWithDescriptor:textureDescriptor
+                                                           iosurface:(__bridge IOSurfaceRef)surface
+                                                               plane:0];
+      FlutterDrawable* drawable = [[FlutterDrawable alloc] initWithTexture:texture
+                                                                     layer:self
+                                                                drawableId:_nextDrawableId++
+                                                                   surface:surface];
+      return drawable;
+    } else {
+      FlutterDrawable* res;
+      CFTimeInterval start = CACurrentMediaTime();
+
+      while (true) {
+        for (FlutterDrawable* drawable in _availableDrawables) {
+          if (!drawable.surface.inUse) {
+            res = drawable;
+            [_availableDrawables removeObject:drawable];
+            goto done;
+          }
+        }
+        usleep(10);
+      }
+    done:
+      CFTimeInterval duration = CACurrentMediaTime() - start;
+      if (duration > 0.003) {
+        NSLog(@"Getting drawable took %f", duration);
+      }
+      return res;
+    }
+  }
+  return nil;
+}
+
+- (void)presentOnMainThread:(FlutterDrawable*)drawable {
```

Comment:
FWIW I tried something like this with the existing drawable/metal layer a few months back. What I found is that there would occassinally be huge amounts of latency.

Does `[CATransaction setDisableActions:YES]` works to prevent the transaction on background thrread state issues? If so, could we not use this to present from a background thread with the regular drawable?

Reply:
I tried that, it helps, but it still not as smooth as no transactions or the approach here. Especially when touch interaction the frame pacing seems quit bad:

<img width="608" alt="Screenshot 2023-11-20 at 18 01 12" src="https://github.com/flutter/engine/assets/96958/f4a6f1ba-f0ae-4cf6-9f22-ce0b1e08e31f">

Reply:
I guess I don't really understand how writing our own Metal layer solves this problem?

Reply:
Not sure about the question? I wasn't able to get `CAMetalLayer` present from background thread with transaction smoothly (that's the first graph - with CATransaction disabled, which helped somewhat, but not completely).

The second graph is `FlutterMetalLayer` (from the PR), in same app, presenting drawables from background thread with CATransaction at consistent 120Hz.

Reply:
But you're dispatching to the main thread to handle the presentation?

Reply:
I am. But we don't really know what `[CAMetalDrawable present]` does. Apparently it's not just setting layer contents to underlying surface, otherwise disabling the transaction would fix it.

Reply:
Right, but then I don't really understand how having our own metal layer helps

Reply:
ack. I might suggest that you find a way to implement this via some opt-in Plist flag or other option that allows for an extended period of testing. But if this solves them problems we've been having with presentation on iOS, I'm all for it!